# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/WebrtcCallWrapper.h
# Commit: e0a2e54e4de8
# Full Hash: e0a2e54e4de80813829cbcf9e4a5e4ff1285d451
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 9
# Description:
#   Bug 1654112 - Move WebRtcCallWrapper to dedicated files. r=ng
#   
#   This patch also renames it WebrtcCallWrapper, to better match neighboring
#   classes starting on "Webrtc".
#   
# ==============================================================================

diff -r 1965757ee924 -r e0a2e54e4de8 dom/media/webrtc/libwebrtcglue/WebrtcCallWrapper.h
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/dom/media/webrtc/libwebrtcglue/WebrtcCallWrapper.h	Thu Apr 08 11:31:43 2021 +0200
@@ -0,0 +1,115 @@
+/* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim:set ts=2 sw=2 sts=2 et cindent: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef DOM_MEDIA_WEBRTC_LIBWEBRTCGLUE_WEBRTCCALLWRAPPER_H_
+#define DOM_MEDIA_WEBRTC_LIBWEBRTCGLUE_WEBRTCCALLWRAPPER_H_
+
+#include <set>
+
+#include "nsISupportsImpl.h"
+
+// libwebrtc includes
+#include "api/video/builtin_video_bitrate_allocator_factory.h"
+#include "call/call.h"
+
+namespace mozilla {
+class MediaSessionConduit;
+
+// Wrap the webrtc.org Call class adding mozilla add/ref support.
+class WebrtcCallWrapper {
+ public:
+  typedef webrtc::Call::Config Config;
+
+  static RefPtr<WebrtcCallWrapper> Create(
+      const dom::RTCStatsTimestampMaker& aTimestampMaker,
+      UniquePtr<media::ShutdownBlockingTicket> aShutdownTicket,
+      SharedWebrtcState* aSharedState, webrtc::WebRtcKeyValueConfig* aTrials);
+
+  static RefPtr<WebrtcCallWrapper> Create(
+      const dom::RTCStatsTimestampMaker& aTimestampMaker,
+      UniquePtr<media::ShutdownBlockingTicket> aShutdownTicket,
+      webrtc::SharedModuleThread* aModuleThread,
+      const webrtc::AudioState::Config& aAudioStateConfig,
+      webrtc::AudioDecoderFactory* aAudioDecoderFactory,
+      webrtc::WebRtcKeyValueConfig* aTrials);
+
+  static RefPtr<WebrtcCallWrapper> Create(UniquePtr<webrtc::Call> aCall);
+
+  NS_INLINE_DECL_THREADSAFE_REFCOUNTING(WebrtcCallWrapper)
+
+  // Don't allow copying/assigning.
+  WebrtcCallWrapper(const WebrtcCallWrapper&) = delete;
+  void operator=(const WebrtcCallWrapper&) = delete;
+
+  webrtc::Call* Call() const;
+
+  bool UnsetRemoteSSRC(uint32_t ssrc);
+
+  // Idempotent.
+  void RegisterConduit(MediaSessionConduit* conduit);
+
+  // Idempotent.
+  void UnregisterConduit(MediaSessionConduit* conduit);
+
+  DOMHighResTimeStamp GetNow() const;
+
+  // Allow destroying the Call instance on the Call worker thread.
+  //
+  // Note that shutdown is blocked until the Call instance is destroyed.
+  //
+  // This CallWrapper is designed to be sharable, and is held by several objects
+  // that are cycle-collectable. TaskQueueWrapper that the Call instances use
+  // for worker threads are based off SharedThreadPools, and will block
+  // xpcom-shutdown-threads until destroyed. The Call instance however will hold
+  // on to its worker threads until destruction.
+  //
+  // If the last ref to this CallWrapper is held to cycle collector shutdown we
+  // end up in a deadlock where cycle collector shutdown is required to destroy
+  // the SharedThreadPool that is blocking xpcom-shutdown-threads from finishing
+  // and triggering cycle collector shutdown.
+  //
+  // It would be nice to have the invariant where this class is immutable to the
+  // degree that mCall is const, but given the above that is not possible.
+  void Destroy();
+
+  const dom::RTCStatsTimestampMaker& GetTimestampMaker() const;
+
+ protected:
+  virtual ~WebrtcCallWrapper();
+
+ private:
+  WebrtcCallWrapper(RefPtr<webrtc::AudioDecoderFactory> aAudioDecoderFactory,
+                    UniquePtr<webrtc::VideoBitrateAllocatorFactory>
+                        aVideoBitrateAllocatorFactory,
+                    UniquePtr<webrtc::Call> aCall,
+                    UniquePtr<webrtc::RtcEventLog> aEventLog,
+                    UniquePtr<webrtc::TaskQueueFactory> aTaskQueueFactory,
+                    const dom::RTCStatsTimestampMaker& aTimestampMaker,
+                    UniquePtr<media::ShutdownBlockingTicket> aShutdownTicket);
+
+  explicit WebrtcCallWrapper(UniquePtr<webrtc::Call> aCall);
+
+  // Allows conduits to know about one another, to avoid remote SSRC
+  // collisions.
+  std::set<MediaSessionConduit*> mConduits;
+  dom::RTCStatsTimestampMaker mTimestampMaker;
+  UniquePtr<media::ShutdownBlockingTicket> mShutdownTicket;
+
+ public:
+  const RefPtr<webrtc::AudioDecoderFactory> mAudioDecoderFactory;
+  const UniquePtr<webrtc::VideoBitrateAllocatorFactory>
+      mVideoBitrateAllocatorFactory;
+  const UniquePtr<webrtc::RtcEventLog> mEventLog;
+  const UniquePtr<webrtc::TaskQueueFactory> mTaskQueueFactory;
+
+ private:
+  // Main thread only, as it's the Call worker thread.
+  UniquePtr<webrtc::Call> mCall;
+};
+
+}  // namespace mozilla
+
+#endif